crates/sncast/src/state/state_file.rs

```diff
@@ -1,13 +1,82 @@
 #![allow(dead_code)]
 use crate::helpers::constants::STATE_FILE_VERSION;
+use crate::response::errors::StarknetCommandError;
 use crate::response::structs::{DeclareResponse, DeployResponse, InvokeResponse};
 use crate::state::hashing::generate_id;
+use crate::WaitForTransactionError;
 use anyhow::{anyhow, Context, Result};
+use cairo_felt::Felt252;
 use camino::Utf8PathBuf;
+use conversions::felt252::SerializeAsFelt252Vec;
 use serde::{Deserialize, Serialize};
 use serde_json::Value;
 use std::collections::HashMap;
 use std::fs;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+struct InnerStateManager {
+    state_file: Utf8PathBuf,
+    executed_transactions_prev_run: ScriptTransactionEntries,
+    executed_transactions_current_run: ScriptTransactionEntries,
+}
+
+#[derive(Default)]
+pub struct StateManager {
+    inner: Option<InnerStateManager>,
+}
+
+impl StateManager {
+    pub fn from(state_file_path: Option<Utf8PathBuf>) -> Result<Self> {
+        let res = if let Some(state_file_path) = state_file_path {
+            let executed_transactions = load_or_create_state_file(&state_file_path)?
+                .transactions
+                .unwrap_or_default();
+
+            Self {
+                inner: Some(InnerStateManager {
+                    state_file: state_file_path,
+                    executed_transactions_prev_run: executed_transactions,
+                    executed_transactions_current_run: ScriptTransactionEntries::default(),
+                }),
+            }
+        } else {
+            Self::default()
+        };
+
+        Ok(res)
+    }
+
+    #[must_use]
+    pub fn maybe_get_success_output(&self, tx_id: &str) -> Option<ScriptTransactionOutput> {
+        if let Some(state) = &self.inner {
+            return state
+                .executed_transactions_prev_run
+                .get_success_output(tx_id);
+        }
+        None
+    }
+
+    pub fn maybe_insert_tx_entry(
+        &mut self,
+        tx_id: &str,
+        selector: &str,
+        result: &Result<impl Into<ScriptTransactionOutput> + Clone, StarknetCommandError>,
+    ) -> Result<()> {
+        if let Some(state) = &mut self.inner {
+            state.executed_transactions_current_run.insert(
+                tx_id,
+                ScriptTransactionEntry::from(selector.to_string(), result),
+            );
+
+            write_txs_to_state_file(
+                &state.state_file,
+                state.executed_transactions_current_run.clone(),
+            )?;
```

Comment:
I am not sure if this is a good idea to overwrite the state file each time after inserting a transaction entry. Would it be possible to do it only once at the end of the script execution?

Reply:
it could be problematic, since the script may panic - then we loose every transaction up until this point. It probably could be done, but I'd be inclined to do this in separate PR

Reply:
Currently if script panics, we end up losing all transactions from the previous execution, which isn't an ideal situation either. I'm fine with creating an issue and addressing it in a separate PR.

Reply:
https://github.com/foundry-rs/starknet-foundry/issues/1967